File: Sources/XMTP/Codecs/TextCodec.swift
//
//  TextCodec.swift
//
//
//  Created by Pat Nakajima on 11/28/22.
//

import Foundation
import XMTPProto

let ContentTypeText = ContentTypeID(authorityID: "xmtp.org", typeID: "text", versionMajor: 1, versionMinor: 0)

enum TextCodecError: Error {
	case invalidEncoding, unknownDecodingError
}

struct TextCodec: ContentCodec {
	typealias T = String

	var contentType = ContentTypeText

	func encode(content: String) throws -> EncodedContent {
		var encodedContent = EncodedContent()

		encodedContent.type = ContentTypeText
		encodedContent.parameters = ["encoding": "UTF-8"]
		encodedContent.content = Data(content.utf8)

		return encodedContent
	}

	func decode(content: EncodedContent) throws -> String {
		if let encoding = content.parameters["encoding"], encoding != "UTF-8" {
			throw TextCodecError.invalidEncoding
		}

		if let contentString = String(data: content.content, encoding: .utf8) {
			return contentString
		} else {
			throw TextCodecError.unknownDecodingError
		}
	}
}

File: library/src/main/java/org/xmtp/android/library/codecs/TextCodec.kt
package org.xmtp.android.library.codecs

import com.google.protobuf.kotlin.toByteStringUtf8
import org.xmtp.android.library.XMTPException

val ContentTypeText = ContentTypeIdBuilder.builderFromAuthorityId(
    "xmtp.org",
    "text",
    versionMajor = 1,
    versionMinor = 0
)

data class TextCodec(override var contentType: ContentTypeId = ContentTypeText) :
    ContentCodec<String> {
    override fun encode(content: String): EncodedContent {
        return EncodedContent.newBuilder().also {
            it.type = ContentTypeText
            it.putAllParameters(mapOf("encoding" to "UTF-8"))
            it.content = content.toByteStringUtf8()
        }.build()
    }

    override fun decode(content: EncodedContent): String {
        val encoding = content.parameters["encoding"]
        if (encoding != null && encoding != "UTF-8") {
            throw XMTPException("Invalid encoding")
        }
        val contentString = content.content.toStringUtf8()
        if (contentString != null) {
            return contentString
        } else {
            throw XMTPException("Unknown decoding")
        }
    }
}
